gadget/gadget.go

```diff
@@ -816,6 +816,13 @@ func validateVolume(vol *Volume, knownFsLabelsPerVolume map[string]map[string]bo
 		var start quantity.Offset
 		if s.Offset != nil {
 			start = *s.Offset
+			// The GPT partition table will start from sector 2 with length
+			// 16384 bytes, end at offset 17407, so if any offset defined
+			// in gadget.yaml is less than 17408, it will break the GPT
+			// partition table and make the built image un-usable.
+			if vol.Schema == schemaGPT && start < 17408 {
```

Comment:
A problem I see here is that sectors can be bigger than 512 bytes, actually we are starting to see 4KiB sectors (the comments should reflect this btw). And we don't know this until we know where we will be flashing the image. So although the check will help in the 512 bytes case, for bigger sectors the problem would go unnoticed - returning an error would be incoherent. Maybe it would be better to print some warning in case the overlap is happening by considering sectors up to 4KiB.

Some unit tests are needed too.

Reply:
I see, I am ok with print some warning messages instead of return error, and may I know what's the syntax for printing a warning message? Sorry I am still pretty new in go lang, what I found so far is `fmt.Printf()`, but I assume that is not for warning message

Here are some code for unit test, mostly copied from the function called `TestValidateVolumeSchema` in `gadget_test.go`, However, currently I think my test might be meaningless because it will always pass if no error will be returned, so I am wondering what is the expectation for this test code?
```go
func (s *gadgetYamlTestSuite) TestValidateVolumeSchemaOffset(c *C) {
        for i, tc := range []struct {
                s   string
                o   quantity.Offset
                err string
        }{
                // Should get a warning message
                {"gpt", 17407, ""},
                // Should be OK
                {"gpt", 17408, ""},
        } {
                c.Logf("tc: %v %+v", i, tc.s)

                err := gadget.ValidateVolume(&gadget.Volume{
                        Name: "name", Schema: tc.s, 
                        Structure: []gadget.VolumeStructure{
                                {Name: "name", Type: "bare", Size: 2048, Offset: &tc.o},
                        },
                }, nil)
                if tc.err != "" {
                        c.Check(err, ErrorMatches, tc.err)
                } else {
                        c.Check(err, IsNil)
                }
        }
}
```

Sorry again that I am pretty new in go lang, and thanks for reviewing this!

Reply:
You could use `logger.Noticef()` to print the warning. I'm not 100% sure if that will print it when using this code from ubuntu-image, but it is worth trying.

And you are indeed right that unit tests cannot check much if we just print something (at least currently we do not have an easy way to do that). We could add some spread test for that, but that is probably an overkill. I suggest instead to revert partially to your first approach returning an error in case there would be an overlap for 512 bytes sectors and printing a warning if the overlap would happen for `512 < sector <= 4 KiB`.

Reply:
I tried to use `logger.Noticef()`, and I will see the following warning message when using ubuntu-image
```
2023/01/09 00:08:43.891151 gadget.go:826: WARNING: offset can not be less than 17408 bytes in gpt schema
```
and when building gadget snap, I won't see any warning message, not sure if this is expected behaviour?

That sounds good to check the sector size first! I have spent some time trying to get the value of SectorSize, but it seems that it is not easy to get the SectorSize when validating the gadget.yaml, do you have any suggestions on this? In the meanwhile, I will still spend some of my leisure time to try, thanks in advance for any help!

Reply:
If you see it when using ubuntu-image, that is good enough, when building it is normal that you don't see it as snapcraft is not using this code.

I was not suggesting to check the sector size, that is quite complicated as you don't know it until you flash the image :slightly_smiling_face:. What I was saying is that we could return an error if the overlap would happen when sectors are 512, which are the smallest ones, but just show a warning if the overlap would happen for sectors up to 4KiB. Note that if the the overlap happens for 512 sectors it would happen also for bigger sectors, but not the other way around.

Reply:
Not sure if this meets your expectation, hopefully I get it right this time :)
```go
                if s.Offset != nil {
                        start = *s.Offset

                        // The GPT partition table will start from sector 2 with length
                        // 16384 bytes
                        if vol.Schema == schemaGPT {
                                if start < 17408 {
                                        // The GPT partition table will end at offset 17407 if sector
                                        // size is 512, so if any offset defined in gadget.yaml is less
                                        // than 17408, it will break the GPT partition table and make
                                        // the built image un-usable.
                                        return fmt.Errorf("invalid structure: offset can not be less than 17408 bytes in gpt schema\n")
                                } else if start < 24576 {
                                        // The GPT partition table will end at offset 24575 if sector
                                        // size is 4096. We are not able to know the sector size when
                                        // building image or gadget snap, so we only print a warning
                                        // message for now
                                        logger.Noticef("WARNING: offset less than 24576 bytes might overlap GPT partition table if sector size is 4K\n")
                                }
                        }
                }
```

Reply:
Yes, something like this is what I meant :)